packages/graphile-build-pg/src/plugins/PgIntrospectionPlugin.js

```diff
@@ -139,287 +138,377 @@ function readFile(filename, encoding) {
+/**
+ * Verifies if the returned PostgresQL structures contain the schemas requested
+ *
+ * @param namespaces
+ * @param requestedSchemas
+ * @param throwIfMissing
+ */
+function verifySchemas(namespaces, requestedSchemas, throwIfMissing) {
+  //extract the schemas found in the database
+  const knownSchemas = namespaces.map(n => n.name);
+
+  //check to see if any of the requested schemas are not actually present in the database
+  const missingSchemas = requestedSchemas.filter(
+    s => knownSchemas.indexOf(s) < 0
+  );
+  if (missingSchemas.length) {
+    const errorMessage = `
+      You requested to use schema '${requestedSchemas.join("', '")}';
+      however we couldn't find some of those! Missing schemas are: '${missingSchemas.join(
+        "', '"
+      )}'
+    `;
+    if (throwIfMissing) {
+      throw new Error(errorMessage);
+    } else {
+      console.warn("⚠️ WARNING⚠️  " + errorMessage); // eslint-disable-line no-console
+    }
+  }
+}
 
-    //extract the schemas found in the database
-    const knownSchemas = introspectionResultsByKind.namespace.map(n => n.name);
-
-    //check to see if any of the requested schemas are not actually present in the database
-    const missingSchemas = schemas.filter(s => knownSchemas.indexOf(s) < 0);
-    if (missingSchemas.length) {
-      const errorMessage = `
-        You requested to use schema '${schemas.join("', '")}';
-        however we couldn't find some of those! Missing schemas are: '${missingSchemas.join("', '")}'
-      `;
-      if (pgThrowOnMissingSchema) {
-        throw new Error(errorMessage);
+/**
+ * Takes the result of getPGStructuresByKind and extracts the tags based on the
+ * smart commenting system exposed by the 'parseTags' function from utils
+ *
+ * @param pgStructuresByKind
+ * @param smartComments - True iff smart comments are enabled
+ */
+function parseTagsFromStructures(pgStructuresByKind, smartComments) {
+  Object.keys(pgStructuresByKind).forEach(kind => {
+    pgStructuresByKind[kind].forEach(struct => {
+      if (smartComments && struct.description) {
+        const parsed = parseTags(struct.description);
+        struct.tags = parsed.tags;
+        struct.description = parsed.text;
       } else {
-        console.warn("⚠️ WARNING⚠️  " + errorMessage); // eslint-disable-line no-console
+        struct.tags = [];
       }
-    }
+    });
+  });
+}
 
+/**
+ * Takes the result of getPGStructuresByKind and creates an index on all of the containing
+ * structures by their ID field; not all structures have an ID field (e.g., attributes), but those that do
+ * will have unique IDs (on a per database basis)
+ *
+ * @param pgStructuresByKind
+ */
+function createStructuresById(pgStructuresByKind) {
+  // put every struct that has an ID field into a single map
+  const pgStructureById = {};
+  Object.keys(pgStructuresByKind).forEach(kind => {
+    pgStructuresByKind[kind].forEach(struct => {
+      if (struct.id) {
+        pgStructureById[struct.id] = struct;
```

Comment:
I cannot find any evidence that object IDs are guaranteed to be unique across the system catalog tables.

The docs state: 

> There are also several alias types for oid: regproc, regprocedure, regoper, regoperator, regclass, regtype, regrole, regnamespace, regconfig, and regdictionary.

and

> The oid type is currently implemented as an unsigned four-byte integer. Therefore, it is not large enough to provide database-wide uniqueness in large databases, or even in large individual tables. So, using a user-created table's OID column as a primary key is discouraged. OIDs are best used only for references to system tables.

> https://www.postgresql.org/docs/10/static/datatype-oid.html

Further pg_depend explicitly states the catalog that the object identifiers it references belong to:

https://www.postgresql.org/docs/10/static/catalog-pg-depend.html

Further, there's no pg table or function that I can find that will tell you what the relevant catalog o find an object in given its object ID.

All of this leads me to believe that you cannot _rely_ on object IDs being globally unique across the system tables, and thus you must factor their type into the lookup (as the existing API does).

Reply:
That's a great catch, and I completely agree. Though now that gets me wondering, since it mentions 

>  only for references to system tables

then that would mean that uniqueness is more or less only guaranteed for references to tables on this list https://www.postgresql.org/docs/current/static/catalogs.html.

Given that, for large databases, it seems possible that there could be oid collisions both globally and within certain specific kinds like PostgreSQL ``types`` for example. That would mean even organizing by id as it is currently set up might be error prone as well?

However, upon looking at the docs for say ``types`` (https://www.postgresql.org/docs/10/static/catalog-pg-type.html), it looks many of the foreign keys are oids that don't belong to system tables. For example, it links to procedures by oid. The only practical way to do this would be for oid uniqueness to be guaranteed within a particular system catalog, right? Even if that isn't specified in the oid datatype docs?

That would lead me to believe that the old way was probably the correct way. Though a possible enhancement might be creating a unique id that is ```oid + calalog_name```.

Reply:
The “kind” is effectively the catalog name, yes 👍

Reply:
Right, so even though it doesn't explicitly say it, the operating assumption is that oid is unique by catalog name (a.k.a. "kind")? Just wanted to double check that was a safe assumption.